# Courses/Tutorials
Garrett Grolemund and Hadley Wickham: [R for Data Science](https://r4ds.had.co.nz/)  
[Illumina DRAGEN Bio-IT Platform](https://www.illumina.com/products/by-type/informatics-products/dragen-bio-it-platform.html)

# Community

## [Biostars](https://www.biostars.org/)
## [Stack Overflow](https://stackoverflow.com/)
## [生信技能树(语雀)](https://www.yuque.com/biotrainee)
## [生信人](https://shengxin.ren/)

# Infrastructure/Core Facility

## [UC Riverside HPCC](https://hpcc.ucr.edu/)
## [中国科学院遗传与发育生物学研究所 基因组分析平台](http://sjzx.genetics.ac.cn/jyzfxpt/ArticleShow.asp?ArticleID=172)

# Databases

## [GTEx: Genotype-Tissue Expression](https://gtexportal.org/home/)

The Genotype-Tissue Expression (GTEx) project is an ongoing effort to build a comprehensive public resource to study **tissue-specific gene expression and regulation**. Samples were collected from **54 non-diseased tissue sites** across **nearly 1000 individuals**, primarily for molecular assays including **WGS**, **WES**, and **RNA-Seq**. Remaining samples are available from the GTEx Biobank. The GTEx Portal provides open access to data including gene expression, QTLs, and histology images.

## [TCGA: The Cancer Genome Atlas](https://portal.gdc.cancer.gov/)

The Cancer Genome Atlas (TCGA), a landmark cancer genomics program, molecularly characterized over 20,000 primary cancer and matched normal samples spanning 33 cancer types. This joint effort between the National Cancer Institute and the National Human Genome Research Institute began in 2006, bringing together researchers from diverse disciplines and multiple institutions.  
Over the next dozen years, TCGA generated over 2.5 petabytes of genomic, epigenomic, transcriptomic, and proteomic data. The data, which has already lead to improvements in our ability to diagnose, treat, and prevent cancer, will remain publicly available for anyone in the research community to use.  

## [ICGC Data Portal](https://dcc.icgc.org/)

The ICGC, established in 2007, aimed to define the genomes of 25,000 primary untreated cancers (the 25K Initiative). The ICGC solved numerous data governance, ethical and logistical challenges to make global genomic data sharing for cancer possible, providing the international community with comprehensive genomic data for many cancer types.  

## [CLLE: Cancer Cell Line Encyclopedia](https://portals.broadinstitute.org/ccle/about)

The CCLE (Cancer Cell Line Encyclopedia) project is a collaboration between the Broad Institute, and the Novartis Institutes for Biomedical Research and its Genomics Institute of the Novartis Research Foundation to conduct a detailed genetic and pharmacologic characterization of a large panel of human cancer models, to develop integrated computational analyses that link distinct pharmacologic vulnerabilities to genomic patterns and to translate cell line integrative genomics into cancer patient stratification. The CCLE provides public access to genomic data, analysis and visualization for over 1100 cell lines.  

## [CrownBio](https://www.crownbio.com/)

## [The Cancer Proteome Atlas](https://www.tcpaportal.org/tcpa/index.html)


## [DICE: Database of Immune Cell eQTL, Expression & Epigenomics](https://dice-database.org/)

TPM

## [Blueprint](http://www.blueprint-epigenome.eu/)

## [FANTOM5: Funcional Annotation of The Mammalian Genome](http://fantom.gsc.riken.jp/5/)

Gene expression data (Count and TPM): http://fantom.gsc.riken.jp/5/datafiles/latest/extra/gene_level_expression/


## [ImmuneCo: ](https://immuco.bjmu.edu.cn/)

## [RefDIC: Reference Database of Immune Cells](http://refdic.rcai.riken.jp/welcome.cgi)

## [IRIS: Immune response in silico](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE22886)
We isolated twelve different types of human leukocytes from peripheral blood and bone marrow, treated them to induce activation and/or differentiation, and profiled their gene expression before and after treatment.  
The twelve cell types are:  
B cells, CD14+ cells, CD4+ CD45RO+ CD45RA- T cells, CD4+ T cells, CD8+ T cells, IgG/IgA memory B cells, IgM memory B cells, Monocytes, NK cells, Neutrophils, Plasma cells from bone marrow, and Plasma cells from PBMC.

## [BioGPS](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE49910)

## [CellO Datasets](http://deweylab.biostat.wisc.edu/cell_type_classification/)
Novel curated dataset comprising nearly all human public, primary bulk samples in the NCBI’s Sequence Read Archive


## [The Human Protein Atlas](https://www.proteinatlas.org/)

## [Roadmap Epigenomics](http://www.roadmapepigenomics.org/)



## [Harmonizome](http://amp.pharm.mssm.edu/Harmonizome/)

## [HCMDB: Human Cancer Metastasis Database](https://hcmdb.i-sanger.com/)

## [AAgAtlas: A human autoantigen database](http://biokb.ncpsb.org/aagatlas/index.php)
## [Open Targets](https://www.opentargets.org/)
## [PreMedKB](http://www.fudan-pgx.org/premedkb/index.html#/home)
## [OncoKB](https://oncokb.org/)
## [GDSC: Genomics of Drug Sensitivity in Cancer](https://www.sanger.ac.uk/science/tools/gdsc-genomics-drug-sensitivity-cancer)
## [GSEA | MSigDB - Broad Institute](http://software.broadinstitute.org/gsea/msigdb/index.jsp)  
## [ImmPort](https://www.immport.org/home)
## [Zenodo](https://zenodo.org/)

## [UpToDate](https://www.uptodate.com/)


# Online Tools

## [DICE Tools](https://tools.dice-database.org/)
## [xCell](http://xcell.ucsf.edu/)
## [CIBERSORT](https://cibersort.stanford.edu/)
## [CIBERSORTx](https://cibersortx.stanford.edu/)
## [ABIS: ABsolute Immune Signal (ABIS) deconvolution](https://github.com/giannimonaco/ABIS)
## [MD Anderson Cancer Center: Public Software](https://bioinformatics.mdanderson.org/public-software/)
## [VDJdb](https://vdjdb.cdr3.net/)





## RNAseq

### Quality Control
* fastp
* FastQC
* AfterQC
* MultiQC

### Trimming
* fastp
* Trimmomatic
* cutadapt


### Alignment
* STAR
* HISAT2

## Assembly


## Quantification
* featureCounts
* HTSeq
* RSME
